### DISCUSSION TITLE
Rust crate

### DISCUSSION THREAD

**Discussion by Napolitain** [uNetworking/uWebSockets#1787] [General]
Hey,

I'm wondering if uWebSockets could be ported at **100% its performance to Rust** (not sure if rewriting in Rust is necessary or if Rust bindings to the C++ library is sufficient). Saw this discussion : https://github.com/uNetworking/uWebSockets/discussions/1593 which  seems to indicate there is not really Rust version currently.

Since uWebSockets.js https://github.com/uNetworking/uWebSockets.js is using the C++ version, I'm wondering :
- how fast is the JS version compared to C++? (I assume, similar?)
- can that approach be easily done for other language like Rust, Go etc ?

Additionally I have a general question on uWS.js : do you think we could "**live at HEAD**" and make uWS.js directly uses at all times the latest uWS release ? I suppose there might be a lot of possible issues which is why you use specific versions ?

Thanks!

**Comment by uNetworkingAB** [comment]
> I'm wondering if uWebSockets could be ported at 100% its performance to Rust

Yes it could. There are initial experiments showing this.

> how fast is the JS version compared to C++? (I assume, similar?)

60% of C++ due to JS wrapping overhead and calling cost for script

> can that approach be easily done for other language like Rust, Go etc ?

Not easily, esp. not for Go.

> do you think we could "live at HEAD" and make uWS.js directly uses at all times the latest uWS release ?

This is already pretty much how it works.
